### ISSUE TITLE
Multi-Threaded downloading for playlists

### ISSUE THREAD

**Issue by JacobHacker** [rg3/youtube-dl#350]
When downloading playlists youtube-dl could use multiple connections to download videos at the same time, right now my dl speed averages at 50k/s, and the full playlist is about 15G (defcon19).

I've been running 4 terminals with --playlist-start=N so far, and spaced them out.

**Comment by singhpratyush** [comment]
I tried may options and here is the best that I came up with - use [aria2](https://github.com/aria2/aria2).

### The `--external-downloader` option
```bash
$ youtube-dl --external-downloader aria2c --external-downloader-args '-x <number of threads>' https://www.youtube.com/playlist?list=PLsPUh22kYmNBkabv9M4fXo6HMLKnc7iR6
```
But this will download only one file at a time. To download many videos simultaneously, you will have to prepopulate download links.

### Prepopulate download links
```bash
$ youtube-dl -g https://www.youtube.com/playlist?list=PLsPUh22kYmNBkabv9M4fXo6HMLKnc7iR6 > dlinks.txt
$ aria2c -i dlinks.txt -j <number of simultaneous downloads> -x <number of threads for each download>
```
You will have to use `>>` instead of `>` in the first command to append more links.

The two commands may also be combined -
```bash
$ aria2c -j <number of simultaneous downloads> -x <number of threads for each download> `youtube-dl -g https://www.youtube.com/playlist?list=PLsPUh22kYmNBkabv9M4fXo6HMLKnc7iR6`
```
But this will result in no backup of your links in case download interrupts.

**Note**: In case of prepopulating, you may end up in a state when your links get expired because your previous downloads would have taken a while.

**Comment by yan12125** [comment]
Let me write some summary. Here are several levels to implement threaded downloading in youtube-dl:

1. User-input URLs, including command line options and batch files. (#109)
2. Playlists (this issue)
3. Segments in segmented streams (HLS, HDS, DASH) (#11044)
4. Chunk level in a single URL (#1498)

Only the last case works with external downloaders like aria2c or axel. For the first three kinds, more codes are necessary.

**Comment by d3fault** [comment]
Multi-Threaded postproc for playlists

Was tempted to open another issue for it, but this is too similar. I'm fine with single threaded downloads (multi threaded downloads introduce the possibility of getting your IP banned), but I'd like to see the postproc done in separate thread(s) (up to N threads, N being the number of CPU cores). As soon as a download is finished, the postproc job should be dispatched and the next download should start immediately. But if all N cores are doing postproc, wait for one to finish before downloading another playlist item.

On my setup postproc is the bottleneck, not bandwidth. Postproc benefits much more from mult-threading, whereas downloading only benefits from multi-threading if the server is throttling you (which implies they don't want you to use multiple connections simultaneously, and might ban you if you do).

**Comment by yan12125** [comment]
@d3fault: Thanks for writing this. Technically, postprocessing and downloading are implemented in different modules, it's better to have an separate issue.

**Comment by DarkTussin** [comment]
One thing you can do is simply write a GUI that runs multiple instances of youtube-dl. That's what I do in the RainDrop downloader - it basically just wraps the youtube-dl process and calls it for as many links as you set. The playlist portion does not run multithreaded however - originally, I wrote the code to pull the links automatically, but since youtube-dl already handled it in a single thread, I left it alone. One thing I've done for personal use is inject some javascript in the youtube playlist browser that just pulls the links into a text box at the bottom (using custom javascript chrome extension). You end up with a list of links and you can just paste them in and watch the multithreaded magic go. You can see it here https://www.youtube.com/watch?v=ihEH4DuTcTg

**Comment by smileytechguy** [comment]
Will this ever be put in the main program?

**Comment by Gaploid** [comment]
Hey Guys, any plans to add multithreaded HLS downloading feature?

**Comment by yaseenwaheed** [comment]
I Need new version

**Comment by crypdick** [comment]
@singhpratyush Is it possible to modify your command to extract audio? In particular, I want to extract audio-- I usually run `youtube-dl  --extract-audio  --format bestaudio --external-downloader aria2c --external-downloader-args '--max-connection-per-server=15' --yes-playlist "<plist_url>"`

**Comment by singhpratyush** [comment]
@crypdick: Yes, it would work fine. You can also do [postprocessing](https://github.com/rg3/youtube-dl#post-processing-options) on the audio if you want, e.g. `--audio-format mp3`.

**Comment by oltodosel** [comment]
my sh snippet
```zsh
ytall() {
	special_o='%(upload_date)s__%(title)s-%(id)s.%(ext)s'
	default_threads=3
	
	if [[ -z $1 ]]; then
		echo "ytall [channel_url/playlist] [threads] [special -o ${special_o}]"
		return
	fi
	
	if [[ -n $3 ]]; then
		youtube-dl -j --flat-playlist $1 | jq -r '.url' | parallel -j $2 --progress youtube-dl -i -o \'$special_o\' -- 
	elif [[ -n $2 ]]; then
		youtube-dl -j --flat-playlist $1 | jq -r '.url' | parallel -j $2 --progress youtube-dl -i --
	else
		youtube-dl -j --flat-playlist $1 | jq -r '.url' | parallel -j $default_threads --progress youtube-dl -i --
	fi
}
```

**Comment by rautamiekka** [comment]
@oltodosel that's a good snippet, but there are 2 things I notice right away:

1) You're limiting the max number of channels/playlists to 1 through the order of the parameters.

2) You're referencing parameter 1 without quoting it.

**Comment by oltodosel** [comment]
@rautamiekka 
1. I don't need it.
2. You're right, guess I got too reliant on oh-my-zsh's automatic url quotation.

**Comment by MPeti1** [comment]
> 
> 
> I tried many options and here is the best that I came up with - use [aria2](https://github.com/aria2/aria2).
> ### The `--external-downloader` option
> 
> ```shell
> $ youtube-dl --external-downloader aria2c --external-downloader-args '-x <number of threads>' https://www.youtube.com/playlist?list=PLsPUh22kYmNBkabv9M4fXo6HMLKnc7iR6
> ```
> 
> But this will download only one file at a time. To download many videos simultaneously, you will have to prepopulate download links.
> ### Prepopulate download links
> 
> ```shell
> $ youtube-dl -g https://www.youtube.com/playlist?list=PLsPUh22kYmNBkabv9M4fXo6HMLKnc7iR6 > dlinks.txt
> $ aria2c -i dlinks.txt -j <number of simultaneous downloads> -x <number of threads for each download>
> ```
> 
> You will have to use `>>` instead of `>` in the first command to append more links.
> 
> The two commands may also be combined -
> 
> ```shell
> $ aria2c -j <number of simultaneous downloads> -x <number of threads for each download> `youtube-dl -g https://www.youtube.com/playlist?list=PLsPUh22kYmNBkabv9M4fXo6HMLKnc7iR6`
> ```
> 
> But this will result in no backup of your links in case download interrupts.
> 
> **Note**: In case of prepopulating links, you may end up in a state when your links get expired because your previous downloads would have taken a while.

There are playlists for which the -g option returns links that start with manifest.googlevideo.com, and if such a link is opened in a browser it will download a few hundred KB file. Is there a solution for cases like that?
Note: your playlist still returns the old links which aria2c can download

**Comment by Pramodh-G** [comment]
Hi, just wanted to say that you can also use GNU parallel to download videos in parallel

Thanks!

**Comment by amadeuszkryze** [comment]
My bash function for parallel: https://github.com/amadeuszkryze/yt-functions

**Comment by Vagelis-Prokopiou** [comment]
I created a multithreaded Rust command line wrapper: https://github.com/Vagelis-Prokopiou/youtube-dl-rs

Currently it serves my own needs, but if it helps anyone else, be my guest and use it.

**Comment by dirkf** [comment]
The comments from 2012 regarding download speed that began this issue are out of date. yt-dl's YouTube download speeds have been fixed, broken and fixed again many times since then.
